File: src/models/account.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::clock::{Clock, SystemClock};

use super::Id;
use super::{IdGenerator, UuidIdGenerator};
use crate::duration::deserialize_duration_opt;

/// Policy for handling balances before the first snapshot.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BalanceBackfillPolicy {
    /// Exclude the account when no balance exists for the date.
    None,
    /// Show the account with a zero value when no balance exists for the date.
    Zero,
    /// Carry back the earliest balance snapshot to earlier dates.
    CarryEarliest,
}

/// An individual financial account (checking, savings, credit card, brokerage, etc.)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    pub id: Id,
    pub name: String,
    pub connection_id: Id,
    #[serde(default)]
    pub tags: Vec<String>,
    pub created_at: DateTime<Utc>,
    pub active: bool,
    /// Opaque data owned by the synchronizer plugin
    #[serde(default, skip_serializing_if = "serde_json::Value::is_null")]
    pub synchronizer_data: serde_json::Value,
}

impl Account {
    pub fn new(name: impl Into<String>, connection_id: Id) -> Self {
        Self::new_with_generator(&UuidIdGenerator, &SystemClock, name, connection_id)
    }

    pub fn new_with(id: Id, created_at: DateTime<Utc>, name: impl Into<String>, connection_id: Id) -> Self {
        Self {
            id,
            name: name.into(),
            connection_id,
            tags: Vec::new(),
            created_at,
            active: true,
            synchronizer_data: serde_json::Value::Null,
        }
    }

    pub fn new_with_generator(
        ids: &dyn IdGenerator,
        clock: &dyn Clock,
        name: impl Into<String>,
        connection_id: Id,
    ) -> Self {
        Self::new_with(ids.new_id(), clock.now(), name, connection_id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::clock::FixedClock;
    use crate::models::{FixedIdGenerator, Id};
    use chrono::TimeZone;

    #[test]
    fn account_new_with_generator_is_deterministic() {
        let fixed_id = Id::from_string("acct-1");
        let ids = FixedIdGenerator::new([fixed_id.clone()]);
        let clock = FixedClock::new(Utc.with_ymd_and_hms(2026, 2, 5, 12, 0, 0).unwrap());

        let account = Account::new_with_generator(&ids, &clock, "Checking", Id::from_string("c"));
        assert_eq!(account.id, fixed_id);
        assert_eq!(account.created_at, clock.now());
    }
}

/// Optional account configuration (stored in account_config.toml).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct AccountConfig {
    /// Override balance staleness for this account.
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_duration_opt"
    )]
    pub balance_staleness: Option<std::time::Duration>,

    /// How to handle portfolio queries before the first balance snapshot.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub balance_backfill: Option<BalanceBackfillPolicy>,
}

File: ts/src/models/account.ts
import { Id } from './id.js';
import { IdGenerator, UuidIdGenerator } from './id-generator.js';
import { Clock, SystemClock } from '../clock.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type BalanceBackfillPolicy = 'none' | 'zero' | 'carry_earliest';

export interface AccountType {
  readonly id: Id;
  readonly name: string;
  readonly connection_id: Id;
  readonly tags: string[];
  readonly created_at: Date;
  readonly active: boolean;
  readonly synchronizer_data: unknown;
}

export interface AccountConfig {
  readonly balance_staleness?: number;
  readonly balance_backfill?: BalanceBackfillPolicy;
  readonly exclude_from_portfolio?: boolean;
}

// ---------------------------------------------------------------------------
// JSON types
// ---------------------------------------------------------------------------

export interface AccountJSON {
  id: string;
  name: string;
  connection_id: string;
  tags: string[];
  created_at: string;
  active: boolean;
  synchronizer_data?: unknown;
}

// ---------------------------------------------------------------------------
// Account namespace (factory functions + serialization)
// ---------------------------------------------------------------------------

export const Account = {
  /**
   * Create a new account with auto-generated id and current time.
   */
  new(name: string, connectionId: Id): AccountType {
    return Account.newWithGenerator(new UuidIdGenerator(), new SystemClock(), name, connectionId);
  },

  /**
   * Create an account with explicit id and timestamp.
   */
  newWith(id: Id, createdAt: Date, name: string, connectionId: Id): AccountType {
    return {
      id,
      name,
      connection_id: connectionId,
      tags: [],
      created_at: createdAt,
      active: true,
      synchronizer_data: null,
    };
  },

  /**
   * Create an account using injectable id generator and clock.
   */
  newWithGenerator(ids: IdGenerator, clock: Clock, name: string, connectionId: Id): AccountType {
    return Account.newWith(ids.newId(), clock.now(), name, connectionId);
  },

  /**
   * Serialize an account to a plain JSON-serializable object.
   * Uses snake_case field names. Omits synchronizer_data when null.
   */
  toJSON(account: AccountType): AccountJSON {
    const json: AccountJSON = {
      id: account.id.toJSON(),
      name: account.name,
      connection_id: account.connection_id.toJSON(),
      tags: [...account.tags],
      created_at: account.created_at.toISOString(),
      active: account.active,
    };
    if (account.synchronizer_data !== null) {
      json.synchronizer_data = account.synchronizer_data;
    }
    return json;
  },

  /**
   * Deserialize an account from a JSON object.
   */
  fromJSON(json: AccountJSON): AccountType {
    return {
      id: Id.fromString(json.id),
      name: json.name,
      connection_id: Id.fromString(json.connection_id),
      tags: [...json.tags],
      created_at: new Date(json.created_at),
      active: json.active,
      synchronizer_data: json.synchronizer_data ?? null,
    };
  },
} as const;
